### PR TITLE
Add context option to configure the indentation of nested nodes for `YamlEncoder`

### LINKED ISSUE #47012: [Serializer] Default Context in Serializer isn't honoured
### Symfony version(s) affected

6.1.2 - 6.13

### Description

The Q&A sums it up perfectly: https://github.com/symfony/symfony/discussions/46117

I noticed it as I was trying to set the below, but noticed it wasn't being honored by the serializer or it's friends:

```yaml
framework:
    serializer:
        default_context:
            skip_null_values: true
```

> # Problem:
> I have been learning Symfony 6 and I got stuck on a weird behaviour of the Serializer component.
> 
> Default instance of Serializer service seems to be ignoring default context set in config file. I am not sure if this is a bug in the framework or rather my misunderstanding how the component and the framework work, but the situation can be easily reproduced in a fresh project.

### How to reproduce

> # Steps to reproduce:
> Environment: **PHP 8.1.4**, **Symfony CLI 5.4.7**, **Symfony 6.0.7**.
> 
> 1. Create a fresh project by executing `symfony local:new --webapp demo`.
> 2. Generate a skeleton console command by executing `bin/console make:command app:demo`.
> 3. Open the generated `config/packages/framework.yaml` and configure Serializer to pretty-print JSON:
> 
> ```yaml
> framework:
>     # ...
>     serializer:
>         default_context:
>             json_encode_options: 128  # value of JSON_PRETTY_PRINT
> ```
> 
> 4. Open the generated `App\Command\DemoCommand` class and request dependency on Serializer in constructor:
> 
> ```
> public function __construct(
>     private \Symfony\Component\Serializer\SerializerInterface $serializer,
> ) {
>     parent::__construct();
> }
> ```
> 
> 5. Modify command's `execute` method to demonstrate the bug:
> 
> ```
> protected function execute(InputInterface $input, OutputInterface $output): int
> {
>     $io = new SymfonyStyle($input, $output);
> 
>     $data = [
>         'foo' => 'bar',
>         'baz' => ['a' => 'b'],
>     ];
>     $context = [
>         // exactly what was configured in default context
>         'json_encode_options' => JSON_PRETTY_PRINT,
>     ];
>     $io->info($this->serializer->serialize($data, 'json')); // this should use default context
>     $io->info($this->serializer->serialize($data, 'json', $context)); // this uses manually configured context
> 
>     return Command::SUCCESS;
> }
> ```
> 
> 6. Execute the command to see the bug:
> 
> ```shell
> $ bin/console -vvv app:wat 
> 
>                                                                                                                         
>  [INFO] {"foo":"bar","baz":{"a":"b"}}                                                                                   
>                                                                                                                         
> 
>                                                                                                                         
>  [INFO] {                                                                                                               
>             "foo": "bar",                                                                                               
>             "baz": {                                                                                                    
>                 "a": "b"                                                                                                
>             }                                                                                                           
>         }                                                                                                               
> ```
> 
> If my understanding of Symfony's documentation is correct, both calls to `serialize` should return the same output. Is this really a bug or did I misunderstood something?

### Possible Solution

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by dbu**
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

Allow to configure the amount of spaces for indentation in Yaml Dumper.

### CODE REVIEW

File: src/Symfony/Component/Serializer/Encoder/JsonEncoder.php [hunk 1]
```diff
@@ -23,17 +23,22 @@ class JsonEncoder implements EncoderInterface, DecoderInterface
...
 
-    public function __construct(JsonEncode $encodingImpl = null, JsonDecode $decodingImpl = null)
+    private $defaultContext;
+
+    public function __construct(JsonEncode $encodingImpl = null, JsonDecode $decodingImpl = null, array $defaultContext = [])
     {
-        $this->encodingImpl = $encodingImpl ?? new JsonEncode();
-        $this->decodingImpl = $decodingImpl ?? new JsonDecode([JsonDecode::ASSOCIATIVE => true]);
+        $this->encodingImpl = $encodingImpl ?? new JsonEncode($defaultContext);
+        $this->decodingImpl = $decodingImpl ?? new JsonDecode(array_merge([JsonDecode::ASSOCIATIVE => true], $defaultContext));
```
Comment (dbu) [inline src/Symfony/Component/Serializer/Encoder/JsonEncoder.php]: or should i even force associative to always be true by switching the order of the arrays in the merge?

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/serializer.php [hunk 2]
```diff
@@ -177,6 +177,7 @@
             ->tag('serializer.encoder')
 
         ->set('serializer.encoder.json', JsonEncoder::class)
+            ->args([null, null])
```
Comment (dbu) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/serializer.php]: why do the tests still fail complaining that arguments 1&2 are not defined?

**Comment by dbu** [comment]
@dunglas does this approach make sense? and would you have an idea what i got wrong with the service configuration?

**Review by maxhelias — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
I've cherry-picked the changes on JsonEncoder in c693778940f379edf42272018cf8984e217befeb, on top of #45554, branch 5.4

I've updated the patch attached to this PR to keep only the part related to YamlEncoder.
This would be a new feature so it should target 6.2

I'll let you update the PR description if you don't mind, also add tests and a line in the changelog of the component?

**Comment by dbu** [comment]
thanks. i rebased on 6.2 and updated the commit.

i think the build errors are not related to my change but broken in the 6.2 branch.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @dbu.

**Pull request merged**